### PR TITLE
feat: Cost Control Center

### PR DESCRIPTION
**Pull request by dreamwing**
Implementation of Cost Control Center frontend and backend as specified in the PRD. Includes diagnostic engine, optimizer service, and dashboard UI.

### CODE REVIEW

File: src/services/optimizer.js [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                if (selectedSkillNames.length === 0) {
+                    throw new Error('No skills selected for removal');
+                }
+
+                const managedSkillsDir = path.join(resolveConfigDir(), 'skills');
+                const allowedSkillNames = new Set();
+                const entries = await fs.readdir(managedSkillsDir, { withFileTypes: true });
+                for (const entry of entries) {
+                    if (!entry.isDirectory() || entry.name.startsWith('.') || entry.name === 'node_modules') continue;
+                    try {
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply src/services/optimizer.js]: Fixed in `bf85e9a`. A04 now mirrors the diagnostics-side skill discovery logic and accepts managed skills that are symlinked directories; I also added a regression test for that case.

File: package.json [hunk 2]
```diff
@@ -9,7 +9,8 @@
     "test:coverage": "jest --forceExit --coverage",
     "lint": "eslint src/ index.js",
     "lint:fix": "eslint src/ index.js --fix",
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply package.json]: Fixed in `96fdd35`. Removed the shared `push` convenience script from `package.json`, so the repo manifest no longer assumes `gh`/`jq` are present for every contributor.

File: public/js/dashboard.js [hunk 3]
```diff
@@ -399,279 +413,786 @@
...
+            method: 'POST',
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify({ backupPath })
+        });
+        if (res.ok) {
+            const result = await res.json();
+            showToast(`✓ Restored ${result.restoredKeys.length} settings from ${result.backupFile}`);
+            await renderHistoryList();
+            // Refresh diagnostics
+            diagnosticsEngine_cache = null; // Assuming this is a global cache variable
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply public/js/dashboard.js]: Fixed in `96fdd35`. Removed the stray `diagnosticsEngine_cache = null` line; `fetchDiagnostics()` is now the only refresh path after undo.

File: public/index.html [hunk 4]
```diff
@@ -257,198 +257,362 @@ <h1>
...
+                                style="background:rgba(255,255,255,0.05); border-radius:20px; padding:6px 14px; display:inline-flex; align-items:center; gap:8px;">
+                                <span style="font-size:11px; color:var(--text-dim);">Est. Monthly</span>
+                                <span style="font-size:12px; font-weight:600; color:var(--accent);"
+                                    id="monthly-forecast">--</span>
+                            </div>
+                            <div
+                                style="background:rgba(255,255,255,0.05); border-radius:20px; padding:6px 14px; display:inline-flex; align-items:center; gap:8px;">
+                                <span style="font-size:11px; color:var(--text-dim);">Cache Hit</span>
+                                <span style="font-size:12px; font-weight:600; color:var(--success);"
+                                    id="cache-hit-rate">--%</span>
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply public/index.html]: Fixed in `96fdd35`. The front-card and optimizer-card cache-hit displays now use distinct DOM ids, and the dashboard JS was updated to target each element explicitly.

File: public/js/dashboard.js [hunk 5]
```diff
@@ -429,288 +437,798 @@
...
+                mainAmount.textContent = '$' + Math.max(0, cur - savingsVal).toFixed(2);
+
+                if (actionsApplied >= totalActions) {
+                    setTimeout(showSuccess, 1000);
                 }
-                return 0;
+            } else {
+                btn.classList.remove('applying');
+                btn.disabled = false;
+                showToast('Optimization failed: ' + (await res.json().catch(() => ({}))).details || 'Unknown error');
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply public/js/dashboard.js]: Fixed in `a7b805e`. The fallback message is now parenthesized correctly, so failed optimization requests show `Unknown error` instead of `Optimization failed: undefined` when the response body has no `details` field.

File: src/services/diagnostics.js [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+                if (!isDir && e.isSymbolicLink()) {
+                    try { isDir = (await fs.stat(fullPath)).isDirectory(); } catch { continue; }
+                }
+                if (!isDir) continue;
+                // Must contain SKILL.md to be a valid Skill
+                try { await fs.access(path.join(fullPath, 'SKILL.md')); } catch { continue; }
+                skillFolders.push({ name: e.name, path: fullPath });
+            }
+
+            const idleDaysThreshold = thresholds.D04_idleDaysThreshold || 7;
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply src/services/diagnostics.js]: Fixed in `82f07eb`. `D04_idleDaysThreshold` now uses nullish coalescing, so an explicit `0` in `diagnostics.config.json` is respected instead of being overridden by the fallback. I also added a regression test covering the zero-threshold case.

File: src/services/optimizer.js [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+        if (result && result.success !== false) {
+            await this.logOptimization({
+                actionId,
+                title: details.title,
+                savings: details.savings,
+                configChanged: details.configChanged,
+                backupPath,
+                undoable: !!backupPath
+            });
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply src/services/optimizer.js]: Fixed in `eb0052b`. `applyAction()` now captures `diagnosticsEngine.runDiagnostics().currentMonthlyCost` before applying the change and passes it through as `preOptCostSnapshot` when logging the optimization, so the 7-day effect tracking path can actually render.

File: public/js/dashboard.js [hunk 8]
```diff
@@ -427,290 +441,1104 @@
...
-                if (show) modal.classList.add('active');
-                else modal.classList.remove('active');
+        }
+
+    // 2. Render Optimization History
+    try {
+        const res = await fetchAuth(API + '/optimizations/history');
+        const history = await res.json();
+        renderHistoryTimeline(document.getElementById('timeline-list'), history);
+        renderHistoryTimeline(document.getElementById('active-timeline-list'), history);
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply public/js/dashboard.js]: Fixed in 8551f60. `renderHistoryList` now checks `res.ok`, surfaces a warning message through the shared error path, and falls back to an empty timeline instead of swallowing a `forEach` type error. I also applied the same `res.ok` guard to `fetchTokens` while touching this path.

File: src/services/pricing.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        const fallbackRatio = 1 - (getNumericPrice(cheapPrice, 'input') / Math.max(getNumericPrice(expPrice, 'input'), Number.EPSILON));
+        return Math.max(0, Math.min(1, parseFloat(fallbackRatio.toFixed(2))));
+    }
+
+    const ratio = 1 - (blendedCheap / blendedExpensive);
+    return Math.max(0, Math.min(1, parseFloat(ratio.toFixed(2))));
+}
+
+async function getReplacements(modelUsage = {}) {
+    if (_replacements) return _replacements;
```
[the remark this answers is not in the document]
Comment (dreamwing) [reply src/services/pricing.js]: Fixed in 8551f60. `diagnosticsEngine.invalidateCache()` now also calls `pricingService.invalidateCache()`, so the cached `_replacements` map is cleared whenever diagnostics state is invalidated after applies or skip-list mutations. That keeps D01 blended savings ratios aligned with fresh model usage data.
